File: 2020/05/part1.js
const fs = require("fs");

Array.prototype.sum = function (callback) {
    return this.reduce(
        (accumulator, element, index) => accumulator + callback(element, index),
        0
    );
};

function parse(str, regexp, callback) {
    return Array.from(str.matchAll(regexp), callback);
}

function parseInteger(str, symbols) {
    return Array.prototype.map
        .call(str, c => symbols.indexOf(c))
        .reverse()
        .sum((digit, position) => digit * symbols.length ** position);
}

const input = fs.readFileSync("input.txt", "utf-8");
const passes = parse(input, /([FB]{7})([LR]{3})/g, match => {
    const row = parseInteger(match[1], "FB"),
        column = parseInteger(match[2], "LR"),
        id = row * 8 + column;
    return { row, column, id };
});

console.log(
    passes.reduce((highest, pass) => (pass.id > highest.id ? pass : highest)).id
);

File: 2021/08/part1.py
import re


def read(filename):
    with open(filename) as file:
        return file.read()


def match(string, pattern, function):
    return [function(*match.groups()) for match in re.finditer(pattern, string)]


def parse(filename, pattern, function):
    return match(read(filename), pattern, function)


data = parse(
    "input.txt", r"([a-z ]+) \| ([a-z ]+)", lambda a, b: (a.split(), b.split())
)

print(
    sum(sum(1 for digit in digits if len(digit) in [2, 3, 4, 7]) for _, digits in data)
)
